feat: new claiming behavior certification on learn

Sembauke opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #43358 
Closes #41040
Closes #40403
<!-- Feel free to add any additional description of changes below this line -->

Sembauke committed:
fix: make setting name an extra step

Sembauke commented:
I am not entirely sure what we agreed on for these issue's, but here is a way to do it.

Sembauke committed:
fix: cypress test

ShaunSHamilton commented on client/src/templates/Introduction/components/CertChallenge.js:
```diff
@@ -89,6 +89,12 @@ const CertChallenge = ({
       )?.show ?? false
     );
 
+    if (!name) {
+      steps.isNameSet = false;
+    } else {
+      steps.isNameSet = true;
+    }
```
```suggestion
    steps.isNameSet = !!name;
```

Sembauke replied:
Is this new syntax?

nhcarrigan replied:
I don't believe it's that new.

ShaunSHamilton replied:
@Sembauke We could always make it more fun:
```js
steps.isNameSet = !!!!!!!!!!!!name;
```

_I wonder if there is a performance penalty for doing that 🤔

Sembauke committed:
Update client/src/templates/Introduction/components/CertChallenge.js

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

naomi-lgbt reviewed (dismissed):
This is looking good. Thanks for doing this!

ShaunSHamilton reviewed (changes requested):
This does not really close either linked issue. Nor should we be mutating the `steps` object within the component that uses it (i.e. adding the `isNameSet` property). Typescript would be complaining, but Redux is not typed yet.

What is needed to close the issues is:
- Change requirements to claim certification (server-side too) to require a Camper's name be set.

What we need clarification on (although, we might be good to go with) is if a Camper has their `name` set to `private`, is their certification hidden? Or, does it default to their username (current behaviour)?

---

My final thought:
- `name` not set -> Cannot claim certificate (needs to be checked on user record during verification)
- `name === private` -> Certificate hidden (cannot be viewed by anyone) -> Flash message explaining it needs to be made public (same as `certifications === private`)

Sembauke commented:
Yeah, I wasn't sure either 👍

Sembauke commented:
I am not entirely sure what you are suggesting here @ShaunSHamilton,

Are you suggesting not adding an extra step and adding a flash message instead?

ShaunSHamilton commented:
> Are you suggesting not adding an extra step and adding a flash message instead?

I am suggesting two options:
1) Remove the `name` step entirely, because it is not a technical requirement to claim a certificate (we just added it, because Campers were getting confused with their username showing up on their cert)
2) What I mentioned above as the needed logic:
> name not set -> Cannot claim certificate (needs to be checked on user record during verification)
> name === private -> Certificate hidden (cannot be viewed by anyone) -> Flash message explaining it needs to be made public (same as certifications === private)

For `2`, essentially, we are enforcing the requirement for `name` to be set and public. Currently, it is not enforced - a Camper can claim a certificate without doing prior.

Hope this is clear. If not, we are probably best hoping on a call with whoever else wants, and discussing for a finality.

Sembauke commented:
![image](https://user-images.githubusercontent.com/46919888/135411976-9f10cc2f-6e8d-4c41-a44e-c219483ccc8e.png)

I get this message after removing the steps for claiming the cert.

Sembauke committed:
fix: use flash message instead

Sembauke committed:
fix: cypress test

nhcarrigan commented on client/i18n/locales/english/translations.json:
```diff
@@ -450,6 +449,7 @@
     "refresh-needed": "You can only use the PaymentRequest button once. Refresh the page to start over.",
     "username-not-found": "We could not find a user with the username \"{{username}}\"",
     "add-name": "This user needs to add their name to their account in order for others to be able to view their certification.",
+    "name-not-public": "This user has their name set to private in order to view their certificaton it needs to be set to public.",
```
```suggestion
    "name-not-public": "This user has their name set to private. In order to view their certification it needs to be set to public.",
```

nhcarrigan commented on client/i18n/locales/english/translations.json:
```diff
@@ -569,7 +569,8 @@
     "complete-project": "Complete {{i18nCertText}} Projects",
     "accept-honesty": "Accept our Academic Honesty Policy",
-    "set-name": "Set your name, and make it public",
+    "set-name-public": "Set your name settings to public",
```
```suggestion
    "set-name-public": "Set your name to be public",
```
tbh I'm not happy with either verbiage but can't think of something better at the moment

Sembauke committed:
Update client/i18n/locales/english/translations.json

Co-authored-by: Nicholas Carrigan (he/him) <nhcarrigan@gmail.com>

Sembauke committed:
fix: remove old keys

moT01 commented:
Sorry if I'm late to the conversation here - I was trying to take a look at this, here's some things I noticed:

In this PR, the cert claiming widget shows 4 steps:
```
Complete the following steps to claim and view your Responsive Web Design Certification
-Accept our Academic Honesty Policy
-Complete Responsive Web Design Certification Projects
-Set your profile settings to public
-Set your certification settings to public
```
If a name is required to claim or view a cert, I think it should be included in the steps. I don't think a name should be required to claim a cert since it hasn't been required in the past and you can just set your name to an empty string afterwards. If it's an empty string, uncheck the step. I think the only things required to claim a cert should be the honesty agreement and the required projects. The step for entering your name can be what it was - `Set your name, and make it public`. If there isn't a name set, the link takes you to the place to add a name - if it is set, it can take you to the privacy setting.

The name should have the same message in the privacy settings as the profile and certification:
`Your certifications will be disabled, if set to private.`

We should not fall back to usernames if a name is private. Names need to be public to view a cert.

My opinion anyway. Seems like that would close both the issues.

Sembauke commented:
Hey @ShaunSHamilton,
I would like to hear your opinion on this above message.

ShaunSHamilton commented:
I suggest we organise a meeting with Quincy, and hash out the future for the privacy settings. Until then, this PR should be blocked.

Sembauke committed:
feat: new behavior

Sembauke committed:
fix: hide unhide settings if account is set to private

Sembauke commented:
I was not at the meeting, but I got a lot from the notes. Let me know what I missed.

Sembauke committed:
feat: update key names

Sembauke committed:
feat: divide share and claim by new button

Sembauke committed:
fix: correct show certificaton button conditons

Sembauke committed:
fix: maybe fix test?

Sembauke committed:
fix: tests possibly

Sembauke committed:
fix: test

Sembauke committed:
fix: remove toggle for buttons

ojeytonwilliams committed:
fix(tests): set all toggles to public

ojeytonwilliams committed:
fix(tests): use privacy toggle command everywhere

This is DRYer and handles the latest changes to /settings

ojeytonwilliams committed:
test: add data-cy links for the 'extra' solution

Sembauke committed:
fix: remove button toggles from test

ojeytonwilliams committed:
fix(tests): give firefox time to toggle settings

Sembauke committed:
fix: wait for profile ui to be updated

ojeytonwilliams committed:
Revert "fix: wait for profile ui to be updated"

This reverts commit 054b0f2a040bf3825c182c53e9a4c13e48f29507.

ojeytonwilliams committed:
fix(tests): clean up project submission

ojeytonwilliams committed:
refactor(tests): DRY out project submission

ojeytonwilliams committed:
refactor: separate privacy toggling and navigation

ojeytonwilliams committed:
fix: remove extraneous visit

I don't know why this ever worked, but if we visit
/learn/responsive-web-design

it *should* scroll to the top of the page, so we were getting lucky that
the test was passing (it should not have)

ShaunSHamilton reviewed (changes requested):
This should not happen (ignore the backlight widget 🤦‍♂️ ):
![image](https://user-images.githubusercontent.com/51722130/137960615-dbf6709f-e530-4818-bf03-3f8c6664889c.png)

The discussed behaviour is
- Campers can **always** view their certs, but a message similar to `profile === private` is given:
![image](https://user-images.githubusercontent.com/51722130/137960801-cd50279f-295c-479b-80e5-7c8307bebf2e.png)

Sembauke committed:
Update client/i18n/locales/english/translations.json

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>
